### PR TITLE
round_out certain cache bounds

### PR DESCRIPTION
**Pull request by jonahwilliams**
This fixes the worst of the pixel jumping issues, though issues with raster cached/animated display list layers are still present.

Fixes https://github.com/flutter/flutter/issues/110959
Fixes https://github.com/flutter/flutter/issues/110957

Previously I computed the bounds using the logical rect, and rounded up to get a texture size. To create the texture for the raster cache entry, we first appended a translation which attempted to position the origin of the display_list/layer at (0,0). This led to a number of issues I did not anticipate, and had difficulty observing on mobile devices with high device pixel ratios.

For display list raster cache entries, this would lead to the raster cached entry being drawn differently than the non-raster cached entry. This is fixed by this PR, by rounding out the bounds.

This does not appear to fix the issue with layer raster caching.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
@dnfield FYI

I anticipate I'll need to update several unittests

File: flow/raster_cache.cc [hunk 1]
```diff
@@ -32,38 +32,19 @@ void RasterCacheResult::draw(SkCanvas& canvas, const SkPaint* paint) const {
...
-  // The image dimensions should always be larger than the device bounds and
-  // smaller than the device bounds plus one pixel, at the same time, we must
-  // introduce epsilon to solve the round-off error. The value of epsilon is
-  // 1/512, which represents half of an AA sample.
-  float epsilon = 1 / 512.0;
-  FML_DCHECK(image_->dimensions().width() - bounds.width() > -epsilon &&
-             image_->dimensions().height() - bounds.height() > -epsilon &&
-             image_->dimensions().width() - bounds.width() < 1 + epsilon &&
-             image_->dimensions().height() - bounds.height() < 1 + epsilon);
+  FML_DCHECK(std::abs(rounded_bounds.size().width() -
```
Comment (jonahwilliams) [inline flow/raster_cache.cc:38]: I think this still might be incorrect. We may have been unconditionally aligning the origin to a physical pixel before, thus the round out would at most add 1 pixel in each direction. Now the the origin can be fractional, rounding out may add up to two pixels in each direction

Comment (jonahwilliams) [reply flow/raster_cache.cc:38]: A potential fix would be to only pixel snap in display_list_layers. That would not have the instability issues that pixel snapping in all composited layers does, but may have other tradeoffs that I'm not aware of

**Comment by jpnurmi** [comment]
I'm seeing this with c130819. A transitioning page gets rasterized at fractional X=-138.6796 and then renders wrong when the transition ends at X=0.0. Things jump in place after moving the mouse cursor over the button to make it re-render without the cache.

https://user-images.githubusercontent.com/140617/188874249-3807abdb-4af6-4438-a50b-567d5594d224.mp4

Screenshot from 2022-09-07 14-08-31

```
RasterCache::Rasterize
	matrix: [  1,0000   0,0000 -138,6796][  0,0000   1,0000   0,0000][  0,0000   0,0000   1,0000]
	logical rect: 0,000000000 0,000000000 960,000000000 633,000000000 (960,000000000x633,000000000)
	dest rect: -139 0 822 633 (961x633)
RasterCacheResult::draw
	matrix: [  1,0000   0,0000 -138,6796][  0,0000   1,0000   0,0000][  0,0000   0,0000   1,0000]
	logical rect: 0,000000000 0,000000000 960,000000000 633,000000000 (960,000000000x633,000000000)
	rounded bounds: -139 0 822 633 (961x633)
RasterCacheResult::draw
	matrix: [  1,0000   0,0000 -25,6627][  0,0000   1,0000   0,0000][  0,0000   0,0000   1,0000]
	logical rect: 0,000000000 0,000000000 960,000000000 633,000000000 (960,000000000x633,000000000)
	rounded bounds: -26 0 935 633 (961x633)
RasterCacheResult::draw
	matrix: [  1,0000   0,0000 -18,1801][  0,0000   1,0000   0,0000][  0,0000   0,0000   1,0000]
	logical rect: 0,000000000 0,000000000 960,000000000 633,000000000 (960,000000000x633,000000000)
	rounded bounds: -19 0 942 633 (961x633)
RasterCacheResult::draw
	matrix: [  1,0000   0,0000  -9,3807][  0,0000   1,0000   0,0000][  0,0000   0,0000   1,0000]
	logical rect: 0,000000000 0,000000000 960,000000000 633,000000000 (960,000000000x633,000000000)
	rounded bounds: -10 0 951 633 (961x633)
RasterCacheResult::draw
	matrix: [  1,0000   0,0000  -4,6098][  0,0000   1,0000   0,0000][  0,0000   0,0000   1,0000]
	logical rect: 0,000000000 0,000000000 960,000000000 633,000000000 (960,000000000x633,000000000)
	rounded bounds: -5 0 956 633 (961x633)
RasterCacheResult::draw
	matrix: [  1,0000   0,0000  -0,8447][  0,0000   1,0000   0,0000][  0,0000   0,0000   1,0000]
	logical rect: 0,000000000 0,000000000 960,000000000 633,000000000 (960,000000000x633,000000000)
	rounded bounds: -1 0 960 633 (961x633)
RasterCacheResult::draw
	matrix: [  1,0000   0,0000   0,0000][  0,0000   1,0000   0,0000][  0,0000   0,0000   1,0000]
	logical rect: 0,000000000 0,000000000 960,000000000 633,000000000 (960,000000000x633,000000000)
	rounded bounds: 0 0 960 633 (960x633)
RasterCacheResult::draw
	matrix: [  1,0000   0,0000   0,0000][  0,0000   1,0000   0,0000][  0,0000   0,0000   1,0000]
	logical rect: 0,000000000 0,000000000 960,000000000 633,000000000 (960,000000000x633,000000000)
	rounded bounds: 0 0 960 633 (960x633)
RasterCacheResult::draw
	matrix: [  1,0000   0,0000   0,0000][  0,0000   1,0000   0,0000][  0,0000   0,0000   1,0000]
	logical rect: 0,000000000 0,000000000 960,000000000 633,000000000 (960,000000000x633,000000000)
	rounded bounds: 0 0 960 633 (960x633)
```

**Comment by jpnurmi** [comment]
The cached image has fractional gaps at the edges because it was rasterized at fractional coordinates. The visual result is off by that fraction when rendered in the origin where the transition ends.

raster_cache_961_633

**Comment by jonahwilliams** [comment]
Yes, I can fix the jumping of display list raster cache entries by performing a round out. Perhaps because there is a round out somewhere in the DL or Skia code (@flar  would be the expert here). What I can't seem to fix is the difference between the layer raster cache entry and non-raster cache entry.

Fixing the jumping is probably more critical, but I still feel like I'm missing a piece of the puzzle here. Simply removing the roundOut for layers seemingly isn't enough

**Comment by jonahwilliams** [comment]
possibly, the issue is something like this: We normally draw the layer at a fractional coordinate, and then the child display list at a fractional coordinate (which gets rounded out to a physical pixel, somewhere).

Now, when we raster cache the layer, we align things such that the top left corner is 0,0. Thus the child display list is already aligned and doesn't move. However, when we draw the layer back on the page, the display list is now at a different position than it would otherwise be. In order to fix this, when raster caching the layer, we need to make sure that the display list ends up at the same fractional coordinate it would have been at, so that whatever is pixel snapping it leaves it at the same position.

**Comment by jonahwilliams** [comment]
but of course, I try fixing that and it doesn't quite work, so maybe I am missing something simple

**Comment by jonahwilliams** [comment]
This latest change forces an integral ctm when rasterizing a layer. That seems to fix the issue where the layers are sometimes rendered on a fractional pixel, but doesn't fix the issue where the layer raster caching is different from the display list raster caching

**Comment by jonahwilliams** [comment]
Okay: last update.

The assumption of the raster cache is that we can replace any layer or display list with a rasterized version of itself. At very low DPR, this is only true if that layer or display list is aligned to physical pixels. Even if we ensure that the texture of the layer or display list is rasterized pixel aligned, when we go to the draw that entry we may find the offset is not physical pixel aligned. Thus we can either snap it to a pixel, leaving it in a different location than the un-rasterized version, or we can draw it in the correct location with obvious aliasing issues.

Unfortunately, that leaves us in a bit of a bind. It means we can't get rid of pixel snapping as long as we have the raster cache. Furthermore, since we didn't notice this issue until the stable release, and there have been many refactors on top of the raster cache since then, it means that a revert is not very straight forward.

Seemingly the only idea long term solution is to revert back to pixel snapping any layers that may be raster cached. The short term change is more uncertain. I can begin working on a patch against the release branch that reverts the removal of pixel snapping. Separately I can also work on re-landing pixel snapping on ToT

**Comment by dnfield** [comment]
@JsouLiang you may find this interesting.

**Comment by flar** [comment]
> Okay: last update.
> 
> The assumption of the raster cache is that we can replace any layer or display list with a rasterized version of itself. At very low DPR, this is only true if that layer or display list is aligned to physical pixels. Even if we ensure that the texture of the layer or display list is rasterized pixel aligned, when we go to the draw that entry we may find the offset is not physical pixel aligned. Thus we can either snap it to a pixel, leaving it in a different location than the un-rasterized version, or we can draw it in the correct location with obvious aliasing issues.
> 
> Unfortunately, that leaves us in a bit of a bind. It means we can't get rid of pixel snapping as long as we have the raster cache. Furthermore, since we didn't notice this issue until the stable release, and there have been many refactors on top of the raster cache since then, it means that a revert is not very straight forward.
> 
> Seemingly the only idea long term solution is to revert back to pixel snapping any layers that may be raster cached. The short term change is more uncertain. I can begin working on a patch against the release branch that reverts the removal of pixel snapping. Separately I can also work on re-landing pixel snapping on ToT

I think this is probably the most accurate assessment. Caching items with fractional translations and displaying them at different fractional translations will have some visual impact. That impact should not be visible much during a transition animation, but for cases where we switch from caching to not caching while the content is relatively unmoving the difference can be detected.

Another option is to always have the endpoints of the rendering be at an integer location which would involve pixel-aware layout. If the before/after locations of a transition were both snapped to a pixel location then the rendering would be consistent between them and for the sub-pixel offsets during an animation from one location to another, the movement would hide the inaccuracies while the animation was happening.

**Comment by jonahwilliams** [comment]
@jpnurmi PTAL at https://github.com/flutter/engine/pull/35981 . Note that you may see pixel snapping after an animated opacity fade in. We'll need to make a separate revert to the framework to fix that.

**Pull request closed without merging**